tools/rewriter/GenCallAsm.cpp

```diff
@@ -481,13 +503,27 @@ static void emit_copy_stack_returns(AsmWriter &aw, size_t stack_return_size, siz
       add_comment_line(aw,
                        "Copy "s + std::to_string(stack_return_size) +
                            " bytes for the return value to the caller's stack");
-      for (int i = 0; i < stack_return_size; i += 8) {
-        add_asm_line(aw, "popq "s + std::to_string(i) + "(%rax)");
+      int i = 0;
+      for (; i + 8 <= stack_return_size; i += 8) {
+        add_asm_line(aw, "movq "s + std::to_string(i) + "(%rsp), %r10");
+        add_asm_line(aw, "movq %r10, "s + std::to_string(i) + "(%rax)");
       }
-
-      if (stack_return_padding > 0) {
-        add_asm_line(aw, "addq $8, %rsp");
+      if (i + 4 <= stack_return_size) {
+        add_asm_line(aw, "movl "s + std::to_string(i) + "(%rsp), %r10d");
+        add_asm_line(aw, "movl %r10d, "s + std::to_string(i) + "(%rax)");
+        i += 4;
+      }
+      if (i + 2 <= stack_return_size) {
+        add_asm_line(aw, "movw "s + std::to_string(i) + "(%rsp), %r10w");
+        add_asm_line(aw, "movw %r10w, "s + std::to_string(i) + "(%rax)");
+        i += 2;
       }
+      if (i < stack_return_size) {
+        add_asm_line(aw, "movb "s + std::to_string(i) + "(%rsp), %r10b");
+        add_asm_line(aw, "movb %r10b, "s + std::to_string(i) + "(%rax)");
+      }
```

Comment:
If the other memcpy gets factored out into its own function this one probably should too.